dot/sync/message.go

```diff
@@ -25,122 +25,154 @@ import (
-	if err != nil {
-		return nil, err
-	}
+		if endNumber.Cmp(bestBlockNumber) == 1 {
+			endNumber = bestBlockNumber
+		}
 
-	endHeader, err := s.blockState.GetHeader(endHash)
-	if err != nil {
-		return nil, err
+		endBlock, err := s.blockState.GetBlockByNumber(endNumber)
+		if err != nil {
+			return nil, err
+		}
+		endHeader = endBlock.Header
+		endHash = endHeader.Hash()
 	}
 
 	logger.Debug("handling BlockRequestMessage", "start", startHeader.Number, "end", endHeader.Number, "startHash", startHash, "endHash", endHash)
 
-	// get sub-chain of block hashes
-	subchain, err := s.blockState.SubChain(startHash, endHash)
-	if err != nil {
-		return nil, err
-	}
-
-	if len(subchain) > int(maxResponseSize) {
-		subchain = subchain[:maxResponseSize]
-	}
-
-	logger.Trace("subchain", "start", subchain[0], "end", subchain[len(subchain)-1])
-
 	responseData := []*types.BlockData{}
 
-	// TODO: check ascending vs descending direction
-	for _, hash := range subchain {
+	// ascending (ie child to parent)
+	if blockRequest.Direction == 0 {
+		for i := endHeader.Number.Int64(); i >= startHeader.Number.Int64(); i-- {
+			blockData, err := s.getBlockData(big.NewInt(i), blockRequest.RequestedData)
+			if err != nil {
+				return nil, err
+			}
+			responseData = append(responseData, blockData)
+		}
+	} else {
+		// descending (ie parent to child)
+		for i := startHeader.Number.Int64(); i <= endHeader.Number.Int64(); i++ {
+			blockData, err := s.getBlockData(big.NewInt(i), blockRequest.RequestedData)
+			if err != nil {
+				return nil, err
+			}
+			responseData = append(responseData, blockData)
+		}
+	}
 
-		blockData := new(types.BlockData)
-		blockData.Hash = hash
+	logger.Debug("sending BlockResponseMessage", "start", startHeader.Number, "end", endHeader.Number)
+	return &network.BlockResponseMessage{
+		BlockData: responseData,
+	}, nil
+}
 
-		// set defaults
-		blockData.Header = optional.NewHeader(false, nil)
-		blockData.Body = optional.NewBody(false, nil)
-		blockData.Receipt = optional.NewBytes(false, nil)
-		blockData.MessageQueue = optional.NewBytes(false, nil)
-		blockData.Justification = optional.NewBytes(false, nil)
+func (s *Service) getBlockData(num *big.Int, requestedData byte) (*types.BlockData, error) {
+	hash, err := s.blockState.GetHashByNumber(num)
+	if err != nil {
+		return nil, err
+	}
 
-		// header
-		if (blockRequest.RequestedData & network.RequestedDataHeader) == 1 {
-			retData, err := s.blockState.GetHeader(hash)
-			if err == nil && retData != nil {
-				blockData.Header = retData.AsOptional()
-			}
+	blockData := new(types.BlockData)
+	blockData.Hash = hash
+	blockData.Header = optional.NewHeader(false, nil)
+	blockData.Body = optional.NewBody(false, nil)
+	blockData.Receipt = optional.NewBytes(false, nil)
+	blockData.MessageQueue = optional.NewBytes(false, nil)
+	blockData.Justification = optional.NewBytes(false, nil)
+
+	if (requestedData & network.RequestedDataHeader) == 1 {
```

Comment:
Can you just return if `requestData == []byte(0)`

Reply:
yeah, will update